### PR TITLE
Warn about unused kwargs in contour methods

### PR DESCRIPTION
**Pull request by dstansby**
I got interested in #1963 (and I think there's another similar bug somewhere out there...), so thought I'd have a go.

The general method behind what I've done is

- Every time a kwarg is retrieved, make sure it is ``pop``ed
- This *should* mean that by the end of the ``__init__`` block, all the kwargs that are used have gone, and any unused kwargs are left behind

I think it should be in theory fairly easy (but very tedious) to roll this out to all methods.

This is probably fairly experimental, so happy for it to be discussed/reviewed and not necessarily merged!

### CODE REVIEW

**Comment by tacaswell** [comment]
I am concerned about spending too much time on this before we drop python2.  Once we are 3 only we can start to use keyword-only arguments so anything the function uses goes in the signature (no popping!) and kwargs are only used when things need to fall through to the next layer down.

File: lib/matplotlib/contour.py [hunk 1]
```diff
@@ -960,6 +961,13 @@ def __init__(self, ax, *args, **kwargs):
 
         self.changed()  # set the colors
 
+        if kwargs:
+            s = ''
+            for key, value in kwargs.items():
+                s += '"' + str(key) + '"' + ', '
+            warnings.warn('The following kwargs were not used by contour: ' +
+                          s)
```
Comment (anntzer) [inline lib/matplotlib/contour.py]: Could just be `", ".join(map(repr, kwargs))`.

**Comment by anntzer** [comment]
Still, it looks like a reasonable improvement.  Minor nitpick, but otherwise looks good.

File: lib/matplotlib/contour.py [hunk 2]
```diff
@@ -1532,7 +1542,7 @@ def _check_xyz(self, args, kwargs):
         Exception class (here and elsewhere).
         """
         x, y = args[:2]
-        self.ax._process_unit_info(xdata=x, ydata=y, kwargs=kwargs)
+        kwargs = self.ax._process_unit_info(xdata=x, ydata=y, kwargs=kwargs)
```
Comment (QuLogic) [inline lib/matplotlib/contour.py]: In this case, it's passed as a mutable value, so you don't technically need to return it. It should be the same object coming out as going in.

Comment (anntzer) [inline lib/matplotlib/contour.py]: I would suggest renaming the method to `_pop_and_process_unit_info` (`_pop_and_apply_unit_info` to save two characters) for clarity -- sneakily modifying a mutable argument usually leads to a lot of head-scratching.

**Comment by tacaswell** [comment]
'powercycled to re-run against current master + restart CI.

**Comment by tacaswell** [comment]
circle CI is failing because it needs to be re-based on current master to pick up the proper config.

**Comment by QuLogic** [comment]
Do we need the rename suggested [here](https://github.com/matplotlib/matplotlib/pull/7728#discussion_r121248223)?

File: lib/matplotlib/contour.py (hunk 2 above)
Comment (tacaswell) [inline lib/matplotlib/contour.py]: I am :-1: on renaming it because it has had the current name for a long time and has always probably had this pop behavior.

In any case that change is outside of the scope if this PR.

**Comment by QuLogic** [comment]
Note that this warning is somewhat the opposite of what's requested in #2369, i.e., pass extra keyword arguments to the artist properties of the generated contour object.
